### PR TITLE
Wrap graphviz diagrams in <div class="graphviz">

### LINKED ISSUE #4865: CSS in default theme breaks clickable map for inheritance-diagrams
Both the default Sphinx theme and the Read-the-Docs theme define

    .rst-content img {max-width: 100%;}

in their `theme.css`. This also gets applied to the diagrams created by the `sphinx.ext.inheritance_diagram` extension, if `graphviz_output_format = 'png'` in `docs/conf.py` (the default). When the diagram is large, the `max-width` CSS setting will effectively scale it down. In this case the clickable map that is overlaid on top of the png image is *not* scaled, resulting in an unusable diagram. This is illustrated in the following screenshot:

screenshot

Note the position of the mouse pointer, which is hovering over the map-location associated with the "KetPlus" box. That click location, however, is not at the visual location of the "KetPlus" box in the diagram.

The problem does not occur for `graphviz_output_format = 'svg'`. A user might work around the problem for the png format by customizing the templates for their documentation by creating the file `docs/_templates/layout.html` to contain

    {% extends "!layout.html" %}
    {% set css_files = css_files + ["_static/mycss.css"] %}

and then define

    img.inheritance {
      max-width: none;
    }

in `docs/_static/mycss.css`.

However, both with this workaround for the png format, and when using the svg format, another problem comes up: The diagrams now extend into the margin, which looks very ugly with most themes. It would be preferable if the diagrams were unscaled, but in a scrollable box (the `overflow: auto` CSS attribute).

Unfortunately, the code generated by the `sphinx.ext.graphviz` plugin (which `sphinx.ext.inheritance_diagram` relies on) creates an `` tag for png or an `<object>` tag for svg, for neither of which the CSS attribute is effective. Therefore, I would suggest that these tags should be enclosed in a `<div class="graphviz">...</div>` to make the diagrams fully stylable via CSS.
Ideally, the standard themes would be modified to both have `max-width: none;`, as above, and

    div.graphviz {
      overflow: auto;
    }

At least, having the additional `<div>` would allow any user to work around the problem in their own custom CSS file.

### PR DESCRIPTION
**Pull request by goerz**
The graphviz extension is modified to wrap the `` and `<object>`
tags it creates for diagrams in the png and svg formats in an additional
`<div class="graphviz">...</div>` tag. This allows to apply additional
CSS styling to the the diagrams that cannot be applied to `` and
`<object>` directly. Most notably,

    div.graphviz {
      overflow: auto;
    }

would be a useful style that the standard themes might consider.

In any case, it would allow users to add custom CSS code to their
project that gets around the problems with inheritance-diagrams
discussed in https://github.com/sphinx-doc/sphinx/issues/4865

The tests for the `grahviz` and `inheritance_diagram` have been modified
to test for the presence of the additional `<div>` tag.

Closes #4865

### CODE REVIEW

**Comment by goerz** [comment]
http://qnet.readthedocs.io/en/latest/symbolic_algebra.html illustrates the result of the applied patch, together with the proper custom CSS to solve #4865: it has scrollable diagrams with proper click-maps

**Comment by goerz** [comment]
Travis reports

    ./sphinx/ext/graphviz.py:301:96: E501 line too long (111 > 95 characters)
    ./sphinx/ext/graphviz.py:306:96: E501 line too long (98 > 95 characters)

Would you like me to fix this?

File: tests/test_ext_graphviz.py
Comment (tk0miya) [inline tests/test_ext_graphviz.py]: Why did you comment out here?

File: sphinx/ext/graphviz.py
Comment (tk0miya) [inline sphinx/ext/graphviz.py]: Why did you fold this line?

File: sphinx/ext/graphviz.py
Comment (tk0miya) [inline sphinx/ext/graphviz.py]: I feel additional `self.body.append` is better:
```
+ self.body.append('<div class="graphviz">')
  self.body.append('<img src="%s" alt="%s" usemap="#%s" %s/>' %
                   (fname, alt, imgmap.id, imgcss))
+ self.body.append('</div>\n')
```

What do you think?

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits.

File: tests/test_ext_graphviz.py
Comment (goerz) [reply tests/test_ext_graphviz.py]: While fixing the tests, I went through the different `assert`s in the tests one by one, and then I forgot to un-comment the last one. Good catch! Fixed.

File: sphinx/ext/graphviz.py
Comment (goerz) [reply sphinx/ext/graphviz.py]: The first time I pushed, TravisCI complained about some overlong lines. In the process of fixing these , I also folded the line in question, even through it wasn't over Travis' 95 character limit. But I suppose you're right, and I shouldn't do "PEP8 cleanups" on lines not directly related to my patch. Reverted.

File: sphinx/ext/graphviz.py
Comment (goerz) [reply sphinx/ext/graphviz.py]: I agree, your solution looks cleaner. I've changed the code to use additional appends. I'll force-push a new version with all of the changes above in a minute.

**Comment by goerz** [comment]
@tk0miya I just force-pushed a new version with all your requested changes. This should be ready to merge now. Thanks for the review!

**Review by tk0miya — APPROVED** [review]
Thank you for quick work.
LGTM!

**Comment by goerz** [comment]
Actually, can you hold off on merging for a minute? I found the tests do not cover the inheritance-diagram with SVG.

Also, for consistency, it would probably best if the SVG `<object>` had the same `class="inheritance"` attribute that the PNG `` has.

**Comment by tk0miya** [comment]
+1 again.
I'll merge this within a week if no objection.

**Comment by tk0miya** [comment]
Merged. Thank you for contribution!

**Pull request merged**
